## Grid Search

In [1]:
import pandas as pd, numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV
digits = load_digits()

In [2]:
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier

In [3]:
model_params = {
    'svm': {
        'model': SVC(),
        'params' : {
            'C': [1, 10, 20],
            'kernel': ['rbf', 'linear'],
            'gamma': ['auto', 'scale'],
        }
    },
    'random_forest': {
        'model': RandomForestClassifier(),
        'params' : {
            'n_estimators': [1, 5, 10, 20, 30, 40]
        }
    },
    'logistic_regression' : {
        'model': LogisticRegression(solver='liblinear', multi_class='auto'),
        'params': {
            'C': [1, 5, 10]
        }
    },
    'GNB' : {
        'model': GaussianNB(),
        'params': {
        }
    },
    'MNB' : {
        'model': MultinomialNB(),
        'params': {
        }
    },
    'decition_tree' : {
        'model': DecisionTreeClassifier(),
        'params': {
            'criterion': ['gini', 'entropy'],
        }
    },
}

In [4]:
scores = []

for model_name, mp in model_params.items():
    clf =  GridSearchCV(mp['model'], mp['params'], cv=10, return_train_score=False)
    clf.fit(digits.data, digits.target)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })
    
df = pd.DataFrame(scores,columns=['model', 'best_score', 'best_params'])
df

model  best_score  \
0                  svm    0.981626   
1        random_forest    0.952132   
2  logistic_regression    0.925975   
3                  GNB    0.811390   
4                  MNB    0.879786   
5        decition_tree    0.827467   

                                    best_params  
0  {'C': 10, 'gamma': 'scale', 'kernel': 'rbf'}  
1                          {'n_estimators': 30}  
2                                      {'C': 1}  
3                                            {}  
4                                            {}  
5                         {'criterion': 'gini'}

#### So, Best model is SVM with C=10, kernel=rbf, gamma=scale (Score : 98.16 %)

In [5]:
model = SVC(C=10, kernel='rbf', gamma='scale')

In [6]:
from sklearn.model_selection import cross_val_score

In [7]:
np.average(cross_val_score(model, digits.data, digits.target, cv=10))

0.9816263190564867

In [8]:
model.fit(digits.data, digits.target)

SVC(C=10)

In [9]:
model.score(digits.data, digits.target)

1.0

In [10]:
model.predict(digits.data[[100]])

array([4])

In [11]:
digits.target[100]

4